TelemetryManager/Src/TMCircularBuffer.cpp

```diff
@@ -1,30 +1,56 @@
 
-#include "CircularBuffer.hpp"
+#include "TMCircularBuffer.hpp"
 
-TMCircularBuffer::TMCircularBuffer(uint8_t* buf, uint16_t size): CircularBuffer(buf, size) {
+TMCircularBuffer::TMCircularBuffer(uint8_t* buf, uint16_t size): CircularBuffer(*buf, size) {
     // Constructor
 }
 
-TMCircularBuffer::TMCircularBuffer(CircularBuffer buf): CircularBuffer(buf) {
+TMCircularBuffer::TMCircularBuffer(CircularBuffer* buf): CircularBuffer(*buf) {
     // Constructor
 }
 
 TMCircularBuffer::~TMCircularBuffer() {
     // Destructor
-}
+    delete[] buf;
+    buf = nullptr;
+ }
+
+TMCircularBuffer::bool TMCircularBuffer::dequeue(MAVLinkByte &byte) {
+    if(getNumAvailBytes() - size_  == 0){
+        return false;
+    else{
+        byte = buf[(readPtr_)];
+        readPtr_ = (readPtr_ +1) % size_;
+        --size_;
+        return true;
+    }
+
 
-TMCircularBuffer::MAVLinkByte TMCircularBuffer::dequeue() { return 0; }
+  }
 
 void TMCircularBuffer::enqueue(MAVLinkByte byte) {
     // Enqueue the byte
+    if(!isFull()){
+        buf_[writePtr_ % size_] = byte;
+        writePtr_++;
+    }
 }
 
-int TMCircularBuffer::lastFullMessageEndIndex() {
+int TMCircularBuffer::bytesUntilLastMessageEnd() {
     /*
-    Rahul: This one is a bit tricky cause you need to know the structure of the MAVLink message. 
+    Rahul: This one is a bit tricky because you need to know the structure of the MAVLink message. 
     I can help you with this one if you want.
     */
+    uint16_t index = readPtr_;
+    uint16_t count = 0;
+    while(count < size_){
+        if(buf_[index] ==0xFD){
+            return index;
+        }
+        index = (index + 1) % size_;
+        count ++;
+    }
     return -1;
 }
 
-int TMCircularBuffer::currentIndex() { return -1; }
+int TMCircularBuffer::currentIndex() { return  readPtr_; }
```

Comment:
Yeah... this won't work, because this class doesn't have the access to the private variables in its parent class